### PR TITLE
threshold_local: n-dimensional support and anisotropic block_size

### PR DESCRIPTION
**Pull request by grlee77**
## Description

`threshold_local` appears to be the only function in `skimage.filters.thresholding` that does not support n-dimensional inputs. I have added n-d support in this PR. This was fairly easy because the underlying `scipy.ndimage` functions used were already n-dimensional.

I was not sure about good 3d test cases, but decided to just replicate the existing 2d test cases via tiling so that the same expected results could be reused.

Allowing anisotropic `block_size` can be useful to enable `block_size = 1` for any non-spatial dimensions as an easy way to batch threshold across multiple channels or timepoints.

## Checklist

- [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- Gallery example in `./doc/examples` (new features only)
- Benchmark in `./benchmarks`, if your changes aren't covered by an
  existing benchmark
- Unit tests
- Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)

## For reviewers

- Check that the PR title is short, concise, and will make sense 1 year
  later.
- Check that new functions are imported in corresponding `__init__.py`.
- Check that new features, API changes, and deprecations are mentioned in
      `doc/release/release_dev.rst`.

### CODE REVIEW

**Review by jni — APPROVED** [review]
Awesome! 🎉

File: skimage/filters/thresholding.py [hunk 1]
```diff
@@ -203,29 +203,37 @@ def threshold_local(image, block_size, method='gaussian', offset=0,
...
+        masks = {}
+        for b in set(block_size):
+            masks[b] = 1. / b * np.ones((b,))
+        # # separation of filters to speedup convolution
+        b = block_size[0]
+        ndi.convolve1d(
+            image, masks[b], axis=0, output=thresh_image, mode=mode, cval=cval)
+        for ax, b in zip(range(1, image.ndim), block_size[1:]):
+            ndi.convolve1d(thresh_image, masks[b], axis=ax,
+                           output=thresh_image, mode=mode, cval=cval)
```
Comment (rfezzani) [inline skimage/filters/thresholding.py:236]:
```suggestion
        kernels = {b: np.full(b, 1/b) for b in set(block_size)}
        # separation of filters to speedup convolution
        thresh_image = image.copy()
        for ax, size in enumerate(block_size):
            ndi.convolve1d(thresh_image, kernels[b], axis=ax,
                           output=thresh_image, mode=mode, cval=cval)
```
Can't we here directly use the [`ndi.uniform_filter`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.ndimage.uniform_filter.html) that is also implemented using the kernel separation?

File: skimage/filters/thresholding.py [hunk 2]
```diff
@@ -203,29 +203,37 @@ def threshold_local(image, block_size, method='gaussian', offset=0,
...
-        mask = 1. / block_size * np.ones((block_size,))
-        # separation of filters to speedup convolution
-        ndi.convolve1d(image, mask, axis=0, output=thresh_image, mode=mode,
-                       cval=cval)
-        ndi.convolve1d(thresh_image, mask, axis=1, output=thresh_image,
-                       mode=mode, cval=cval)
+        masks = {}
+        for b in set(block_size):
+            masks[b] = 1. / b * np.ones((b,))
+        # # separation of filters to speedup convolution
```
Comment (mkcor) [inline skimage/filters/thresholding.py:230]:
```suggestion
        # separation of filters to speed up convolution
```

File: skimage/filters/thresholding.py (hunk 1 above)
Comment (grlee77) [reply skimage/filters/thresholding.py:236]: yes, I we can just make a single call to `uniform_filter`. thanks

File: skimage/filters/thresholding.py (hunk 2 above)
Comment (grlee77) [reply skimage/filters/thresholding.py]: good suggestion, but this line ended up being removed altogether when switching from `convolve1d` to `uniform_filter`

**Review by rfezzani — APPROVED** [review]
Thank you @grlee77 :tada:

**Pull request merged**